x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenizer.java

```diff
@@ -0,0 +1,447 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.nlp.tokenizers;
+
+import com.carrotsearch.hppc.CharObjectHashMap;
+import com.carrotsearch.hppc.CharObjectMap;
+
+import org.apache.lucene.analysis.CharArrayMap;
+import org.apache.lucene.analysis.CharArraySet;
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.OffsetAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+import org.elasticsearch.common.Strings;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class BpeTokenizer extends Tokenizer {
+    private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+    private final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
+    private final PositionIncrementAttribute posIncAtt = addAttribute(PositionIncrementAttribute.class);
+
+    private static final Pattern pattern = Pattern.compile(
+        "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)|\\s+"
+    );
+
+    static Map<Integer, Character> byteEncoder() {
+        List<Integer> bytes = IntStream.concat(
+            IntStream.range(Character.codePointAt("!", 0), Character.codePointAt("~", 0) + 1),
+            IntStream.concat(
+                IntStream.range(Character.codePointAt("¡", 0), Character.codePointAt("¬", 0) + 1),
+                IntStream.range(Character.codePointAt("®", 0), Character.codePointAt("ÿ", 0) + 1)
+            )
+        ).boxed().collect(Collectors.toList());
+        List<Integer> chars = new ArrayList<>(bytes);
+        int n = 0;
+        for (int i = 0; i < 2 << 7; i++) {
+            if (bytes.contains(i)) {
+                continue;
+            }
+            bytes.add(i);
+            chars.add((2 << 7) + n);
+            n++;
+        }
+        char[] charArray = new char[chars.size()];
+        int i = 0;
+        for (Integer c : chars) {
+            charArray[i++] = Character.toChars(c)[0];
+        }
+        Map<Integer, Character> map = new HashMap<>(charArray.length, 1.0f);
```

Comment:
Why is this using a load factor of 1?  That will make lookups slower because there will likely be collisions.

Another data structure to consider would be an array of size 256 where the key is the index into the array.  That should also be faster because instead of multiple key comparisons and an unnecessarily generic memory layout the lookup is just calculating a single memory offset.

Reply:
> Another data structure to consider would be an array of size 256 where the key is the index into the array. That should also be faster because instead of multiple key comparisons and an unnecessarily generic memory layout the lookup is just calculating a single memory offset.

Yeah, since we 100% know the chars && ints and they will never change, we should just create state associative arrays and be done with it.

Reply:
Ah, well, we still need a hashmap for the reverse look up, but a straight up `char[]` will work for the encoder just fine :)